src/Storages/System/StorageSystemFilesystemCache.cpp

```diff
@@ -0,0 +1,58 @@
+#include "StorageSystemFilesystemCache.h"
+#include <DataTypes/DataTypeString.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <DataTypes/DataTypeTuple.h>
+#include <Common/FileCache.h>
+#include <Common/FileCacheFactory.h>
+#include <Interpreters/Context.h>
+#include <Disks/IDisk.h>
+
+
+namespace DB
+{
+
+NamesAndTypesList StorageSystemFilesystemCache::getNamesAndTypes()
+{
+    return {
+        {"cache_base_path", std::make_shared<DataTypeString>()},
+        {"cache_path", std::make_shared<DataTypeString>()},
+        {"file_segment_range_begin", std::make_shared<DataTypeUInt64>()},
+        {"file_segment_range_end", std::make_shared<DataTypeUInt64>()},
+        {"size", std::make_shared<DataTypeUInt64>()},
+        {"state", std::make_shared<DataTypeString>()},
+        {"cache_hits", std::make_shared<DataTypeUInt64>()},
+        {"references", std::make_shared<DataTypeUInt64>()},
+    };
+}
+
+StorageSystemFilesystemCache::StorageSystemFilesystemCache(const StorageID & table_id_)
+    : IStorageSystemOneBlock(table_id_)
+{
+}
+
+void StorageSystemFilesystemCache::fillData(MutableColumns & res_columns, ContextPtr, const SelectQueryInfo &) const
+{
+    auto caches = FileCacheFactory::instance().getAll();
+
+    for (const auto & [cache_base_path, cache_data] : caches)
+    {
+        const auto & cache = cache_data.cache;
+        auto holder = cache->getAll();
```

Comment:
This method needs to be removed, it is not correct to hold pointers to segments here, because no one can do that apart from CachedReadBufferFromRemoteFS because otherwise cache file will fail to be removed when file is removed in IDIsk interface.
Instead better to have a method like `getSnapshotState`